src/header/common/range.rs

```diff
@@ -0,0 +1,185 @@
+use std::fmt::{self, Display};
+use std::str::FromStr;
+
+use header::{Header, HeaderFormat, RangeUnit};
+use header::parsing;
+
+/// The `Range` header defined by HTTP/1.0.
+///
+/// > The "Range" header field on a GET request modifies the method
+/// > semantics to request transfer of only one or more subranges of the
+/// > selected representation data, rather than the entire selected
+/// > representation data.
+///
+/// > Range = byte-ranges-specifier / other-ranges-specifier
+/// > other-ranges-specifier = other-range-unit "=" other-range-set
+/// > other-range-set = 1*VCHAR
+///
+/// # Example values
+/// * `bytes=1000-`
+/// * `bytes=-2000`
+/// * `bytes=0-1,30-40`
+/// * `custom_unit=0-123`
+///
+/// Spec: https://tools.ietf.org/html/rfc7233#section-3.1
+#[derive(PartialEq, Clone, Debug)]
+pub struct Range {
+    unit: RangeUnit,
+    ranges: Vec<RangeSpec>,
+}
+
+/// Each 'Range' header can contain one or more RangeSpecs.
+/// Each RangeSpec defines a range of units to fetch
+#[derive(PartialEq, Clone, Debug)]
+pub enum RangeSpec {
+    /// Get all bytes between x and y ("x-y")
+    FromTo(u64, u64),
+    /// Get all bytes starting from x ("x-")
+    AllFrom(u64),
+    /// Get last x bytes ("-x")
+    Last(u64)
+}
+
+impl Range {
+    /// Get the most common byte range header ("bytes=from-to")
+    pub fn simple(from: u64, to: u64) -> Range {
+        Range {
+            unit: RangeUnit::Bytes,
+            ranges: vec![RangeSpec::FromTo(from, to)],
+        }
+    }
+}
+
+
+impl fmt::Display for RangeSpec {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            RangeSpec::FromTo(from, to) => write!(fmt, "{}-{}", from, to),
+            RangeSpec::Last(pos) => write!(fmt, "-{}", pos),
+            RangeSpec::AllFrom(pos) => write!(fmt, "{}-", pos),
+        }
+    }
+}
+
+
+impl fmt::Display for Range {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        let as_strings : Vec<String> = self.ranges.iter().map(ToString::to_string).collect();
```

Comment:
This will make several heap allocations, before being pushed into the `fmt` buffer and then being tossed. It'd be faster to skip those. For instance:

```rust
try!(write!(fmt, "{}=", self.unit));

let mut iter = self.ranges.iter();
match iter.next() {
    Some(range) => try!(write!("{}", range)),
    None => return Ok(())
}
for range in iter {
    try!(write!(",{}", range));
}
Ok(())
```